app/src/modules/battery/battery.c

```diff
@@ -0,0 +1,174 @@
+/*
+ * Copyright (c) 2024 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#include <zephyr/kernel.h>
+#include <zephyr/logging/log.h>
+#include <zephyr/zbus/zbus.h>
+#include <zephyr/drivers/sensor/npm1300_charger.h>
+#include "nrf_fuel_gauge.h"
+#include <math.h>
+#include <zephyr/sys/util.h>
+
+#include "lp803448_model.h"
+#include "message_channel.h"
+
+/* Register log module */
+LOG_MODULE_REGISTER(battery, CONFIG_APP_BATTERY_LOG_LEVEL);
+
+/* Register subscriber */
+ZBUS_SUBSCRIBER_DEFINE(battery, CONFIG_APP_BATTERY_MESSAGE_QUEUE_SIZE);
+
+#define FG_BURST_COUNT 10
+#define FG_BURST_DELAY K_MSEC(100)
+
+static const struct device *charger = DEVICE_DT_GET(DT_NODELABEL(npm1300_charger));
+static int64_t fg_ref_time;
+static float max_charge_current;
+static float term_charge_current;
+static float fg_burst_current_accumulated = 0;
+static float fg_burst_delta_accumulated = 0;
+static int fg_burst_index = 0;
+
+static int charger_read_sensors(float *voltage, float *current, float *temp)
+{
+	struct sensor_value value;
+	int ret;
+
+	ret = sensor_sample_fetch(charger);
+	if (ret < 0) {
+		return ret;
+	}
+
+	sensor_channel_get(charger, SENSOR_CHAN_GAUGE_VOLTAGE, &value);
+	*voltage = (float)value.val1 + ((float)value.val2 / 1000000);
+
+	sensor_channel_get(charger, SENSOR_CHAN_GAUGE_TEMP, &value);
+	*temp = (float)value.val1 + ((float)value.val2 / 1000000);
+
+	sensor_channel_get(charger, SENSOR_CHAN_GAUGE_AVG_CURRENT, &value);
+	*current = (float)value.val1 + ((float)value.val2 / 1000000);
+
+	return 0;
+}
+
+static void sample(void)
+{
+	/* Fuel gauge variables */
+	float voltage;
+	float current;
+	float temp;
+	float state_of_charge;
+	float delta;
+	float time_to_empty;
+	float time_to_full;
+	static struct nrf_fuel_gauge_state_info fg_state;
+	int ret;
+
+	/* Update fuel gauge state */
+	ret = charger_read_sensors(&voltage, &current, &temp);
+	if (ret) {
+		LOG_ERR("Error: Could not read from charger device: %d", ret);
+		SEND_FATAL_ERROR();
+		return;
+	}
+
+	delta = (float) k_uptime_delta(&fg_ref_time) / 1000.f;
+	fg_ref_time = k_uptime_get();
+	state_of_charge = nrf_fuel_gauge_process(voltage, current, temp, delta, &fg_state);
+
+	/* calculate TTE/TTF */
+	time_to_empty = nrf_fuel_gauge_tte_get();
+	time_to_full = nrf_fuel_gauge_ttf_get((bool)-max_charge_current, -term_charge_current);
+
+	fg_burst_current_accumulated += delta * current;
+	fg_burst_delta_accumulated += delta;
+
+	/* Repeat readings upto FG_BURST_COUNT measurement until we collect enough data points */
+	if (fg_burst_index++ < FG_BURST_COUNT) {
```

Comment:
I don't fully understand this logic. Why not just return without calling `sample()` first? If it related to the intended zbus event flow to subscribers? It's a good idea to run the fuel gauge for some iterations before transmitting the data. It might take some time for the state to stabilize. Especially TTF and TTE will be quite off in the beginning

Reply:
Result after offline discussion: Will remove this logic and see if it is sufficient enough for our use case.